### PR TITLE
start testing Enzyme

### PR DESCRIPTION
**Pull request by CarloLucibello**
We don't officially support Enzyme yet, but since we are going to in the future we better start testing, see what is missing, and watch out for regressions.
The testing work has already started at https://github.com/EnzymeAD/Enzyme.jl/issues/805, here I mostly copy some of the tests contributed by @jgreener64 so that they become part of Flux's testsuite.

Related issues in Enzyme.jl:
-  https://github.com/EnzymeAD/Enzyme.jl/issues/805
- https://github.com/EnzymeAD/Enzyme.jl/issues/1261
- https://github.com/EnzymeAD/Enzyme.jl/issues/1347

PRs in CUDA.jl:
- https://github.com/JuliaGPU/CUDA.jl/pull/1869

### CODE REVIEW

File: test/ext_enzyme/enzyme.jl [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    loss(model, x) = sum(model(x))
+    
+    @testset "RNN" begin
+        model = RNN(3 => 5)
+        x = randn(Float32, 3, 10)
+        Flux.reset!(model)
+        grads_flux = Flux.gradient(loss, model, x)
+        Flux.reset!(model)
+        grads_enzyme = grad(loss, model, x)
+        check_grad(grads_flux[1].state, grads_enzyme[1].state, broken=true)
```
Comment (CarloLucibello) [inline test/ext_enzyme/enzyme.jl:127]: For this we have 
```julia
ulia> grads_flux[1].state
5×1 Matrix{Float32}:
 -7.479321
  0.1483739
  3.8861375
  6.9063663
  3.9330518

julia> grads_enzyme[1].state
5×10 Matrix{Float32}:
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

Comment (CarloLucibello) [reply test/ext_enzyme/enzyme.jl:127]: I think doesn't make sense at all to test the state

Comment (darsnack) [reply test/ext_enzyme/enzyme.jl:127]: You mean as opposed to testing w.r.t. the weights? We can do both, but I think testing w.r.t. state is necessary. It's unique in that it is a mutable field, changes shape, and actually needs gradients to flow through it for BPTT (vs. normalization state which does not).

**Comment by CarloLucibello** [comment]
if tests pass as they should this can be merged.

This PR gives us already some good cpu coverage. For CUDA instead, I get some pretty fundamental problems at the moment https://github.com/EnzymeAD/Enzyme.jl/issues/1347.  GPU tests will be added in future PRs

**Comment by CarloLucibello** [comment]
ready to merge

**Comment by CarloLucibello** [comment]
bump

**Review by mcabbott — APPROVED** [review]

**Pull request merged**